src/lambdaconnect_model/transformations.cljc

```diff
@@ -4,29 +4,63 @@
+  (into {} (concat 
+            (map (fn [[k a]] [k [(t/datomic-name a) (parser-for-attribute a) (:optional a)]]) (:attributes entity))
+            (map (fn [[k r]] [k [(t/datomic-name r) (parser-for-relationship r) (:optional r)]]) (:relationships entity))
+            [[:entity-hash (hash entity)]
+             [:attribute-parser-hash (hash parser-for-attribute)]
+             [:relationship-parser-hash (hash parser-for-relationship)]])))
+
+(def parser-cache (atom {}))
+
 (defn json-to-clojure
   "Converts json-based map into the one that conforms to the spec. "
-  [json entity parser-for-attribute parser-for-relationship]
-  (rebuild-map json 
-               (fn [k val]
-                 (let [key (clojure.string/replace k #"_" "-")
-                       attr (get (:attributes entity) key)
-                       rel  (get (:relationships entity) key)]
-                   (assert (or rel attr) (str "The json does not match the entity '" (:name entity) "'. The attribute attribute: " key " with value: '" val "' does not exist in the model."))
-                   [(t/datomic-name (or attr rel))
-                    (if (and (nil? val) (:optional attr)) nil
-                        ((if attr (parser-for-attribute attr) (parser-for-relationship rel)) val))]))))
+  ([json entity parser-for-attribute parser-for-relationship]
+   (let [cached (get @parser-cache (:name entity))
+         parsers (if (and cached 
+                          (= (:attribute-parser-hash cached) (hash parser-for-attribute))
+                          (= (:relationship-parser-hash cached) (hash parser-for-relationship))
+                          (= (:entity-hash cached) (hash entity)))
+                   cached
+                   (let [built (build-parser-cache entity parser-for-attribute parser-for-relationship)]
+                     (swap! parser-cache #(assoc % (:name entity) built))
+                     built))]
+     (rebuild-map json 
+                  (fn [key val]
+                    (let [[datomic-name parser optional?] (get parsers key)]
+                      (assert parser (str "The json does not match the entity '" (:name entity) "'. The attribute attribute: " key " with value: '" val "' does not exist in the model."))
+                      [datomic-name
+                       (if (and (nil? val) optional?) 
+                         nil
+                         (parser val))]))))))
 
-(defn- inverses-fun [names entity]
-  (into {} (map (fn [n] (let [inverse (t/relationship-for-inverse-name entity n)]
-                          [n (when inverse (t/datomic-name inverse))])) names)))
+(defn- inverses-fun [entity]
+  (let [non-datomic-rels (difference (set (vals (:relationships entity)))
+                                     (set (vals (:datomic-relationships entity))))]
+    (assoc 
+    (->> non-datomic-rels
+         (map #(-> [(t/datomic-inverse-name %) (t/datomic-name %)]))
```

Comment:
Faktycznie, może to wygląda lepiej niż vector, chociaż nie jestem przekonany...
W tym konkretnym przypadku z kolei napisałbym `(juxt t/datomic-inverse-name t/datomic-name)`, z jakiegoś powodu polubiłem `juxt`, ale nie musisz tego zmieniać, jeśli nie chcesz